database/src/main/java/com/firebase/ui/database/FirebaseArray.java

```diff
@@ -26,69 +28,72 @@
  * This class implements an array-like collection on top of a Firebase location.
  */
 class FirebaseArray implements ChildEventListener {
-    public interface OnChangedListener {
+    public interface ChangeEventListener {
         enum EventType {ADDED, CHANGED, REMOVED, MOVED}
 
         void onChanged(EventType type, int index, int oldIndex);
 
-        void onCancelled(DatabaseError databaseError);
+        void onCancelled(DatabaseError error);
     }
 
+    protected ChangeEventListener mListener;
+    protected boolean mIsListening;
     private Query mQuery;
-    private OnChangedListener mListener;
     private List<DataSnapshot> mSnapshots = new ArrayList<>();
 
     public FirebaseArray(Query ref) {
         mQuery = ref;
+    }
+
+    public void setChangeEventListener(@NonNull ChangeEventListener listener) {
+        if (mIsListening && listener == null) {
+            throw new IllegalStateException("Listener cannot be null.");
+        }
+        mListener = listener;
+    }
+
+    public void startListening() {
+        if (mListener == null) throw new IllegalStateException("Listener cannot be null.");
         mQuery.addChildEventListener(this);
+        mIsListening = true;
     }
 
-    public void cleanup() {
+    public void stopListening() {
         mQuery.removeEventListener(this);
+        mSnapshots.clear();
+        mIsListening = false;
     }
 
-    public int getCount() {
+    public int size() {
```

Comment:
Woah woah woah! 😄 I was just saying that eventually @puf will have to open source it (technically he doesn't, but come on!). I think he was waiting for some additional refinement and this PR + another one I have ready will clean things up a bit, making it more likely that `FirebaseArray` will be open sourced. Then yes, you will finally be able to implement a proper `FirebaseSpinnerAdapter`!